nixos/modules/services/backup/syncoid.nix

```diff
@@ -170,47 +181,56 @@ in {
     # Implementation
 
     config = mkIf cfg.enable {
-      users =  {
+      users = {
         users = mkIf (cfg.user == "syncoid") {
           syncoid = {
             group = cfg.group;
             isSystemUser = true;
+            # For syncoid to be able to create /var/lib/syncoid/.ssh/
+            # and to use custom ssh_config or known_hosts.
+            home = "/var/lib/syncoid";
+            createHome = true;
           };
         };
         groups = mkIf (cfg.group == "syncoid") {
           syncoid = {};
         };
       };
 
-      systemd.services.syncoid = {
-        description = "Syncoid ZFS synchronization service";
-        script = concatMapStringsSep "\n" (c: lib.escapeShellArgs
-          ([ "${pkgs.sanoid}/bin/syncoid" ]
-            ++ (optionals c.useCommonArgs cfg.commonArgs)
-            ++ (optional c.recursive "-r")
-            ++ (optionals (c.sshKey != null) [ "--sshkey" c.sshKey ])
-            ++ c.extraArgs
-            ++ [ "--sendoptions" c.sendOptions
-                 "--recvoptions" c.recvOptions
-                 "--no-privilege-elevation"
-                 c.source c.target
-               ])) (attrValues cfg.commands);
-        after = [ "zfs.target" ];
-        serviceConfig = {
-          ExecStartPre = (map (pool: lib.escapeShellArgs [
-            "+/run/booted-system/sw/bin/zfs" "allow"
-            cfg.user "hold,send" pool
-          ]) (getPools "source")) ++
-          (map (pool: lib.escapeShellArgs [
-            "+/run/booted-system/sw/bin/zfs" "allow"
-            cfg.user "create,mount,receive,rollback" pool
-          ]) (getPools "target"));
-          User = cfg.user;
-          Group = cfg.group;
-        };
-        startAt = cfg.interval;
-      };
+      systemd.services = mapAttrs' (name: c:
+        nameValuePair "syncoid-${escapeUnitName name}" (mkMerge [
+          {
+            description = "Syncoid ZFS synchronization from ${c.source} to ${c.target}";
+            script = lib.escapeShellArgs ([ "${pkgs.sanoid}/bin/syncoid" ]
+              ++ (optionals c.useCommonArgs cfg.commonArgs)
+              ++ (optional c.recursive "-r")
+              ++ (optionals (c.sshKey != null) [ "--sshkey" c.sshKey ])
+              ++ c.extraArgs
+              ++ [ "--sendoptions" c.sendOptions
+                   "--recvoptions" c.recvOptions
+                   "--no-privilege-elevation"
+                   c.source c.target
+                 ]);
+            after = [ "zfs.target" ];
+            serviceConfig.ExecStartPre =
+              map (pool: lib.escapeShellArgs [
+                "+/run/booted-system/sw/bin/zfs" "allow"
+                cfg.user "hold,send" pool
+              ]) (localPoolName c.source) ++
+              map (pool: lib.escapeShellArgs [
+                "+/run/booted-system/sw/bin/zfs" "allow"
+                cfg.user "create,mount,receive,rollback" pool
+              ]) (localPoolName c.target);
+            serviceConfig.User = cfg.user;
```

Comment:
Indeed, see also https://github.com/jimsalterjrs/sanoid/issues/532#issuecomment-697159696.
I've set `PrivateTmp=true` and also taken the time to find working values for the security options, based upon `systemd-analyze security syncoid-$NAME.service` and previous experiences like https://github.com/NixOS/nixpkgs/issues/98904 .
As always, it might still be too much hardening. Please bear with me.

Reply:
The hardening doesn't cause any problems for me, although as always it is definitely an increased maintenance burden that requires in depth systemd knowledge to understand. I'd like to get someone else's opinion, although it seems to be nearly impossible to find reviewers for sanoid/syncoid PRs for some reason.

Reply:
I'm trying out sanoid/syncoid, hopefully I'll have the experience in a few weeks to have actual opinions. I'll try to remember this (and other PRs) -- but feel free to ping me if it's still stuck in a while.

Reply:
Thanks, it would be great to get this merged. Make sure you use https://github.com/NixOS/nixpkgs/pull/83904, otherwise sanoid templates won't work right.